File: Έγγραφα SEM4105/Λύσεις ασκήσεων/05. Πέμπτο εργαστήριο/3/determinant.c
#include <stdio.h>
#include <math.h>



// max abs value of a(row,:)
double maxval(int n, double const a[], int row)
{
  int mj = 0;
    
  for (int j = 1; j < n; ++j) {
    if (fabs(a[row+j*n]) > fabs(a[row+mj*n])) {
      mj=j;
    }
  }

  return fabs(a[row+mj*n]);
}

void swap(double * a, double *b)
{
  double t = *a;
  *a = *b;
  *b = t;
}

void pivot(int k, int n, double a[],int * change)
{
  int mi = k;
  double mvmi = maxval(n,a,mi);
  for (int i = k+1; i < n; ++i) {
    double  mvi = maxval(n,a,i);
    
    if (fabs(a[i+k*n] / mvi) > fabs(a[mi+k*n] / mvmi) ) {
      mi = i;
      mvmi = mvi;
    }
  }

  if (mi!=k) {
    for (int j=k; j < n; ++j) {
      swap(&a[mi+j*n], &a[k+j*n]);
    }
    ++*change;
  }
  
}



void triang(int n, double a[], int * change)
{
  for (int k = 0; k < n-1; ++k) {

    pivot(k,n,a,change);
	
    for (int i = k+1; i < n; ++i) {
      double const g = -a[i+n*k] / a[k+n*k];
      for (int j = k; j < n; ++j) {
	a[i+j*n] += g * a[k +j*n];
      }
    }
  }
}






double det(int n, double a[])
{
  int change = 0;
  
  triang(n,a,&change);

  double d = 1.0;
  for (int i=0; i < n; ++i) {
    d *= a[i+i*n];
  }

  return (change%2 == 0? d : -d);
}


int
main()
{
  int const n=4;
  double a[n*n];

  a[0+n*0] = 2.1;
  a[0+n*1] = 3.9;
  a[0+n*2] = 0.3;
  a[0+n*3] = -4.1;
  a[1+n*0] = 4.3;
  a[1+n*1] = -1.3;
  a[1+n*2] = 0.8;
  a[1+n*3] = 1.5;
  a[2+n*0] = 1.0;
  a[2+n*1] = -2.8;
  a[2+n*2] = 4.3;
  a[2+n*3] = -8.1;
  a[3+n*0] = 2.4;
  a[3+n*1] = 6.1;
  a[3+n*2] = -1.1;
  a[3+n*3] = 12.5;

  printf("%g\n", det(n,a));
}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/05. Πέμπτο εργαστήριο/3/determinant.f90
! a<->b
ELEMENTAL SUBROUTINE swap(a,b)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (inout) :: a, b

  DOUBLE PRECISION :: temp

  temp = a
  a    = b
  b    = temp

END SUBROUTINE swap


!     MAKE LOWER TRIANGULAR
SUBROUTINE TRIANG(A, change)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (inout) :: A(:,:)
  INTEGER, INTENT (out) :: change

  INTEGER :: N, I, K, M
  DOUBLE PRECISION :: G
  DOUBLE PRECISION, PARAMETER :: eps = 1d-7

  INTERFACE 
     ELEMENTAL SUBROUTINE swap(a,b)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (inout) :: a, b
     END SUBROUTINE swap
  END INTERFACE

  N = SIZE(A,1)

  change = 0
  DO K=1, N-1
     M = MAXLOC(ABS(A(K:N,K)),1) + K-1
     
     IF (M /= K) THEN
        CALL swap(A(K,:), A(M,:))
        change = change + 1
     ENDIF
     
     DO I=K+1, N
        G = -A(I,K) / A(K,K)
        A(I,K:N) = A(I,K:N) + A(K,K:N) * G
     ENDDO
  ENDDO

END SUBROUTINE TRIANG

! gives the determinant of a
FUNCTION det(a)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (in) :: a(:,:)
  DOUBLE PRECISION :: det

  DOUBLE PRECISION :: b(SIZE(a,1),SIZE(a,2))
  INTEGER :: i, change

  INTERFACE
     SUBROUTINE triang(a, change)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (inout) :: a(:,:)
       INTEGER, INTENT (out) :: change
     END SUBROUTINE triang
  END INTERFACE

  b = a
  CALL triang(b, change) ! now b is lower triangular

  det = 1.0d0
  DO i = 1,SIZE(a,1)
     det = det * b(i,i)
  ENDDO

  IF (MOD(change, 2) == 1) det = -det
END FUNCTION det


PROGRAM deter
  IMPLICIT NONE

  INTEGER, PARAMETER :: n = 4
  DOUBLE PRECISION :: a(n,n)

  INTERFACE
     FUNCTION det(a)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (in) :: a(:,:)
       DOUBLE PRECISION :: det
     END FUNCTION det
  END INTERFACE

  A(1,1) = 2.1D0
  A(1,2) = 3.9D0
  A(1,3) = 0.3D0
  A(1,4) = -4.1D0

  A(2,1) = 4.3D0
  A(2,2) = -1.3D0
  A(2,3) = 0.8D0
  A(2,4) = 1.5D0

  A(3,1) = 1.0D0
  A(3,2) = -2.8D0
  A(3,3) = 4.3D0
  A(3,4) = -8.1D0

  A(4,1) = 2.4D0
  A(4,2) = 6.1D0
  A(4,3) = -1.1D0
  A(4,4) = 12.5D0

  PRINT *, det(a)

END PROGRAM deter
